test/com/google/javascript/jscomp/Es6RewriteClassExtendsExpressionsTest.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2018 The Closure Compiler Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.javascript.jscomp;
+
+import com.google.javascript.jscomp.CompilerOptions.LanguageMode;
+
+public final class Es6RewriteClassExtendsExpressionsTest extends CompilerTestCase {
+
+  @Override
+  protected CompilerPass getProcessor(Compiler compiler) {
+    return new Es6RewriteClassExtendsExpressions(compiler);
+  }
+
+  @Override
+  protected void setUp() throws Exception {
+    super.setUp();
+    setAcceptedLanguage(LanguageMode.ECMASCRIPT_NEXT);
+    setLanguageOut(LanguageMode.ECMASCRIPT3);
+    disableTypeCheck();
+    enableRunTypeCheckAfterProcessing();
+  }
+
+  public void testBasic() {
+    test(
+        "const foo = {'bar': Object}; class Foo extends foo['bar'] {}",
+        lines(
+            "const foo = {'bar': Object};",
+            "const testcode$classextends$var0 = foo['bar'];",
+            "class Foo extends testcode$classextends$var0 {}"));
+  }
+
+  public void testName() {
+    testSame("class Foo extends Object {}");
+  }
+
+  public void testGetProp() {
+    testSame("const foo = { bar: Object}; class Foo extends foo.bar {}");
+  }
+
+  public void testMixinFunction() {
+    test(
+        lines(
+            "/** @return {function(new:Object)} */",
+            "function mixObject(Superclass) {",
+            "  return class extends Superclass {",
+            "    bar() { return 'bar'; }",
+            "  };",
+            "}",
+            "class Foo {}",
+            "class Bar extends mixObject(Foo) {}"),
```

Comment:
@ChadKillingsworth How would you tell Closure here that `Bar` should be augmented with the interface of the internal class? The way we've typically been doing that is to manually define an interface corresponding to the inner mixin class, and then assert that the base class `@implements` it:

```
/** @interface */
MixinInterface = function(){};
/** @return {string} */
MixinInterface.prototype.bar = function(){};

/**
 * @constructor
 * @extends {SuperClass}
 * @implements {MixinInterface}
 */
const base = Mixin(SuperClass);

class Foo extends base {
}
```

My understanding is that if you were to put the `@implements` annotation directly on the `Foo` class, Closure would require that `Foo` implements it, rather than trusting that it is implemented as in the separate `@constructor` case.

Reply:
This isn't fully supported yet - this is just step one. This allows the compiler to successfully transpile this code. Type checking is yet to come.